### ISSUE TITLE
DbCommand byte code instrumentation

### ISSUE THREAD

**Issue by Kielek** [open-telemetry/opentelemetry-dotnet-instrumentation#1126]
# Feature Request

**Are you requesting automatic instrumentation for a framework or library? Please describe.**

DbCommand intrumentation.
It should cover also `Oracle.ManagedDataAccess.dll` library requested in #1117  but it cover more libraries not only Oracle.

**Describe the solution you'd like**
Cover all DbCommand based libraries.

**Describe alternatives you've considered**
Have dedicated instrumentations for specific libraries as in upstream side: https://github.com/DataDog/dd-trace-dotnet/tree/master/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AdoNet

**Additional context**
Verify how it will be working with Npgsql. We have source code instrumentation there.

**Comment by muhaook** [comment]
Hi @Kielek , during my prototype, I noticed System.Data got loaded in shared appdomain on .NET framework. 
```
[2022-08-21T15:54:07.404708400Z] [33532|35180] [debug] Comparing signature for method: System.Data.Common.DbCommand..ctor
[2022-08-21T15:54:07.404737400Z] [33532|35180] [debug] Enqueue for ReJIT [ModuleId=140712098334424, MethodDef=06002de7, AppDomainId=140713692092832, IsDomainNeutral=1, Assembly=System.Data, Type=System.Data.Common.DbCommand, Method=.ctor, Signature=200001]
[2022-08-21T15:54:07.404761800Z] [33532|35180] [debug] ModuleLoadFinished stored metadata for 140712098334424 System.Data AppDomain 140713692092832 EE Shared Assembly Repository
```
I remember the current bytecode instrumentation has a limitation on assembly in shared appdomain. Could you please confirm?

**Comment by muhaook** [comment]
@pjanotti , as far as I know, below means it is a shared appdomain
```
[2022-08-21T15:54:07.404761800Z] [33532|35180] [debug] ModuleLoadFinished stored metadata for 140712098334424 System.Data AppDomain 140713692092832 EE Shared Assembly Repository
```
Attached please find the full logs
[dotnet-tracer-native-w3wp-33532.log](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/files/9437649/dotnet-tracer-native-w3wp-33532.log)

**Comment by pjanotti** [comment]
@muhaook thanks for the logs. I think the root of the problem is: 
```terminal
[2022-08-21T15:54:05.429461400Z] [33532|36148] [info] AssemblyLoadFinished: OpenTelemetry.AutoInstrumentation.dll was not loaded domain-neutral
```

We still have to understand why this is happening in this case...

**Comment by pjanotti** [comment]
Hi @muhaook, this was fixed, see #1134 and respective PR #1234. In order for the instrumentation assembly (`OpenTelemetry.AutoInstrumentation.dll`) to be loaded as domain neutral you need to registry it in the GAC.

I'm going to close this issue. Let's know if you hit other problems.

**Comment by pjanotti** [comment]
@RassK @pellared adding/removing `OpenTelemetry.AutoInstrumentation.dll` to the GAC via the setup script seems a good idea (see previous comment).

**Comment by pellared** [comment]
@pjanotti I am not sure if this issue should be closed. The issue reported by @muhaook was reported (and closed) here: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1117

This issue is about adding generic bytecode instrumentation for `System.Data.Common.DbCommand`.

**Comment by pjanotti** [comment]
Makes sense @pellared I was thinking about failing to load as domain-neutral and that is fixed, but, the issue is to track adding DbCommand. Now it is unblocked, one can write it :)

**Comment by pjanotti** [comment]
Facilitating installation of required assemblies in the GAC is tracked via #1823

**Comment by Kielek** [comment]
Removing vNext, Splunk would like to work on this withing 3-4 months timeframe. Hopefully 1.15.0 target.

Q: How to handle scenarios where both DbCommand instrumentation is executed together with SqlClient/Oracle?
Should we make it opt-in only as in file based configuration, or just make a note that you can expect more than one span related to the same query?

**Comment by nrcventura** [comment]
> Q: How to handle scenarios where both DbCommand instrumentation is executed together with SqlClient/Oracle?
> Should we make it opt-in only as in file based configuration, or just make a note that you can expect more than one span related to the same query?

Even if we follow an opt-in model, it may be possible that an application needs to interact with multiple datastores which could lead double instrumentation. We may want to see if there's a way to suppress instrumentation such as the case with grpc where you can have both grpc and http spans created.
